megamek/src/megamek/client/Client.java

```diff
@@ -1125,7 +1115,85 @@ public String receiveReport(Vector<Report> v) {
         for (Report r : v) {
             report.append(r.getText());
         }
-        return report.toString();
+
+        Set<Integer> set = new HashSet<Integer>();
+        //find id stored in spans and extract it
+        Pattern p = Pattern.compile("<s(.*?)n>");
+        Matcher m = p.matcher(report.toString());
+
+        //add all instances to a hashset to prevent duplicates
+        while(m.find()){
+            String cleanedText = m.group(1).replaceAll("\\D", "");
+            if(cleanedText.length() > 0) {
+                set.add(Integer.parseInt(cleanedText));
+            }
+        }
+
+        String updatedReport = report.toString();
+        //loop through the hashset of unique ids and replace the ids with img tags
+        for (int i : set) {
+            if(getCachedImageData(i) != null) {
+                updatedReport = updatedReport.replace("<span id='" +i + "'></span>", getCachedImageData(i));
+            }
+        }
+        return updatedReport;
+    }
+
+    /**
+     * get img tag for unit id
+     */
+    private String getCachedImageData(int id){
+        if(imgCache == null || !imgCache.containsKey(id)) {
+            return null;
+        }
+        return imgCache.get(id);
+    }
+
+    /**
+     * Hashtable for storing image tags containing base64Text src
+     */
+    private void cacheImageData(Entity entity){
```

Comment:
I'm fairly certain that the TilesetManager has its own caching mechanism. I don't think you need to have a separate caching mechanism here.

Reply:
I looked at that, but I'm storing strings after the image is encoded as a Base64 string and put into an \<img\> tag. the data looks something like the following, but just quite a bit longer. 

`<img src='data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADgAAAAwCAYAAABE1blzAAALvkl...'>`

I do actually use TilesetManager's image cache when I call `imageFor` though. Perhaps I should rename `cacheImageData` so that it doesn't sound like I'm just recaching the images?

Reply:
Now that I look at it again, I get what you're doing.